src/coreclr/jit/forwardsub.cpp

```diff
@@ -667,10 +661,23 @@ bool Compiler::fgForwardSubStatement(Statement* stmt)
     //
     // We may sometimes lose or change a type handle. Avoid substituting if so.
     //
-    if (gtGetStructHandleIfPresent(fwdSubNode) != gtGetStructHandleIfPresent(fsv.GetNode()))
+    // However, we allow free substitution of hardware SIMD types.
+    //
+    CORINFO_CLASS_HANDLE fwdHnd = gtGetStructHandleIfPresent(fwdSubNode);
+    CORINFO_CLASS_HANDLE useHnd = gtGetStructHandleIfPresent(fsv.GetNode());
+    if (fwdHnd != useHnd)
     {
-        JITDUMP(" would change struct handle (substitution)\n");
-        return false;
+        if ((fwdHnd == NO_CLASS_HANDLE) || (useHnd == NO_CLASS_HANDLE))
+        {
+            JITDUMP(" would add/remove struct handle (substitution)\n");
+            return false;
+        }
+
+        if (!isHWSIMDClass(fwdHnd) || !isHWSIMDClass(useHnd))
+        {
+            JITDUMP(" would change struct handle (substitution)\n");
+            return false;
+        }
```

Comment:
What is the reason to disallow this in the first place?

Reply:
I don't know if Andy ran into some other issues when bringing up forward sub, but the primary reason for exact handle match I would think of is to avoid substituting into call arguments.

This is also the reason we cannot allow more SIMD types here.

Reply:
How relevant is the first reason with recent call arg work? I would assume we're not far off from being able to let call args morphing handle this on its own.

Reply:
> How relevant is the first reason with recent call arg work?

It basically comes down to replacing this:

https://github.com/dotnet/runtime/blob/fae7ee8e7e3aa7f86836318a10ed676641e813ad/src/coreclr/jit/morph.cpp#L2350

With `arg.GetSignatureClassHandle`.

(And also deleting some other oddities like `InferOpSizeAlign` using `lvStructDoubleAlign`, `GetHfaType(argx`), etc)

Reply:
> there are also (now rare, or actually non-existent, I did not check) cases where block copies can end up with sizeof(LHS) != sizeof(RHS).

Could you elaborate on how this is impactful? For `TYP_SIMD` and any non-`TYP_STRUCT`, I'd expect we treat a block copy as just a raw copy of the bits and so the handle shouldn't matter.

Reply:
I don't remember the details, but all these exclusions were added because without them things failed downstream.

Could be more of these can be relaxed.

Reply:
> Could you elaborate on how this is impactful?

Imagine we have a mismatched copy like this (we definitely used to have those before #69271, I am not sure if some cases still exist):
```c
struct<8> a = struct<16> b;
Use(a);
```
Substituting `a` with `b` could lead to all sorts of interesting things.

Reply:
I would expect that one to not get a forward sub since the `varType` should differ, let alone the class handle.

That is, `TYP_SIMD8` shouldn't be substitutable for a `TYP_SIMD16`. Just `TYP_SIMD16` for any other `TYP_SIMD16`, regardless of `simdJitBaseType`

Reply:
> I would expect that one to not get a forward sub since the `varType` should differ

In my example they're both `TYP_STRUCT`, so that check would pass.

Reply:
I'd expect for `TYP_STRUCT` we have to check the handles and for that case to fail since they mismatch.

I'd expect that for `TYP_STRUCT` and `TYP_REF` that checking the handle is needed. I'd expect for most primitive types and simd types, just the var type is enough.

There may of course be nuance around the integer types since `TYP_UINT` often appears as a `TYP_INT` with "unsignedness" being tracked elsewhere.

Reply:
In the "ideal future", we will need no handle checks whatsoever, because the "types" would implicitly match at the point of the `ASG`. Currently, we need the handle checks for `varTypeIsStruct` nodes for the reasons outlined above.